planner: introduce a new variable `tidb_regard_null_as_point` (#8157)
pingcap/docs-cn: system-variables.md

- 默认情况下（即该变量值为 `ON` 时），某张表上的统计信息过期后，优化器认为该表上除总行数以外的统计信息不再可靠，转而使用 pseudo 统计信息。将该变量值设为 `OFF` 后，即使统计信息过期，优化器也仍会使用该表上的统计信息。
- 如果表数据修改较频繁，没有及时对表执行 `ANALYZE`，但又希望执行计划保持稳定，可以将该变量值设为 `OFF`。

> ### `tidb_regard_null_as_point` <span class="version-mark">从 v5.4.0 版本开始引入</span>

Reviewer: The system variables are sorted by name in this doc.

- 作用域：SESSION | GLOBAL
- 默认值：`ON`

> - 这个变量用来控制优化器是否可以把 null 值当做点值并作为前缀条件来访问索引。

Reviewer 2: 请教下点值具体是什么值呢？搜了下 TiDB 文档，没出现过这个词

Reviewer 3: 这里的点值跟优化器在索引上构造 scan range 有关。对于每个 range 来说，只有最后一列可以不为点值，而前面所有列都必须为点值。比如一段 range 为 `[a1 a2 ... an, b1 b2 ... bn]`，那么必然有 `a1 = b1`, ..., `a{n-1} = b{n-1}`，即前 n - 1 列都是点值。 在下面这个例子中，对于多列索引 `indx(a, b)`，如果 `tidb_regard_null_as_point` 为 `OFF`，构造的 range 是 `[null, null]`。如果 `tidb_regard_null_as_point` 为 `ON`，构造的 range 是 `[null 1, null 1]`。

Author: `这个变量用来控制优化器是否可以把包含 null 的等值条件作为前缀条件来访问索引` 这样呢 @qiancai  改成 “包含 null 的等值条件"

> - 比如在有多列索引 `index(a, b)` 及查询条件 `a<=>null and b=1` 的情况下，如果关闭开关，则不能把 `b=1` 用来进行索引访问，因为 null 值不能被当做固定前缀。

Reviewer 2: ```suggestion
- 比如在有多列索引 `index(a, b)` 且查询条件中某列为空 `a<=>null and b=1` 的情况下，如果关闭该变量，则不能把 `b=1` 用来进行索引访问，因为 null 值不能被当做固定前缀。
```

Reviewer 3: 某列为空是指 `a<=>null` 吗

- 这个变量用于控制在记录 TiDB 日志和慢日志时，是否将 SQL 中的用户信息遮蔽。
- 将该变量设置为 `1` 即开启后，假设执行的 SQL 为 `insert into t values (1,2)`，在日志中记录的 SQL 会是 `insert into t values (?,?)`，即用户输入的信息被遮蔽。

### `tidb_regard_null_as_point` <span class="version-mark">从 v5.4.0 版本开始引入</span>

- 作用域：SESSION | GLOBAL
- 默认值：`ON`
- 这个变量用来控制优化器是否可以将包含 null 的等值条件作为前缀条件来访问索引。

> - 该变量默认开启，能减少需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1`包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。

Reviewer 2: ```suggestion
- 该变量默认开启，能减少需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1`包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。
```

> - 该变量默认开启，能减少需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1`包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。

Reviewer 2: ```suggestion
- 该变量默认开启，能减少需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。
```

> - 该变量默认开启，能减少需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。

Reviewer 2: ```suggestion
- 该变量默认开启，能减少优化器需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。
```

> - 该变量默认开启，能减少优化器需要访问的索引数据量，提高查询执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 值的等值条件，优化器不会使用 `b=1` 进行索引访问。

Reviewer 4: ```suggestion
- 该变量默认开启。开启后，优化器需要访问的索引数据量减少，查询的执行速度提高。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 的等值条件，优化器不会使用 `b=1` 进行索引访问。
```

> - 该变量默认开启。开启后，该变量可以减少优化器需要访问的索引数据量，从而提高查询的执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 的等值条件，优化器不会使用 `b=1` 进行索引访问。

Reviewer 2: ```suggestion
- 该变量默认开启。开启后，该变量可以使优化器减少需要访问的索引数据量，从而提高查询的执行速度。例如，在有多列索引 `index(a, b)` 且查询条件为 `a<=>null and b=1` 的情况下，优化器可以同时使用查询条件中的 `a<=>null` 和 `b=1` 进行索引访问。如果关闭该变量，因为 `a<=>null and b=1` 包含 null 的等值条件，优化器不会使用 `b=1` 进行索引访问。
```
